earth_enterprise/src/fusion/portableglobe/rewritedbroot_unittest.cpp

```diff
@@ -0,0 +1,166 @@
+  ASSERT_FALSE(dbroot.nested_feature(4).has_kml_url());
+}
+
+TEST(ReplaceReferencedKmlTest, NestedEntryPreserveTest) {
+  const string FILE_1 = "kmlfile.kmz";
+  const string FILE_2 = "newfile.kml";
+  const string BEFORE_URL_1 = "http://server.name/path_a/" + FILE_1;
+  const string BEFORE_URL_2 = "http://other.server.name/path_b/" + FILE_2;
+  const string AFTER_BASE_URL = "https://different.server/path_c";
+  const string AFTER_URL_1 = AFTER_BASE_URL + "/" + FILE_1;
+  const string AFTER_URL_2 = AFTER_BASE_URL + "/" + FILE_2;
+  geProtoDbroot dbroot;
+  
+  dbroot.add_nested_feature();
+  dbroot.add_nested_feature();
+  addKmlUrlAsValue(BEFORE_URL_1, &dbroot);
+  dbroot.add_nested_feature();
+  addKmlUrlAsValue(BEFORE_URL_2, &dbroot);
+
+  ReplaceReferencedKml(AFTER_BASE_URL, "", true, &dbroot);
+  ASSERT_FALSE(dbroot.nested_feature(0).has_kml_url());
+  ASSERT_FALSE(dbroot.nested_feature(1).has_kml_url());
+  ASSERT_EQ(dbroot.nested_feature(2).kml_url().value(), AFTER_URL_1);
+  ASSERT_FALSE(dbroot.nested_feature(3).has_kml_url());
+  ASSERT_EQ(dbroot.nested_feature(4).kml_url().value(), AFTER_URL_2);
+}
+
+keyhole::dbroot::StringEntryProto * addTranslationEntry(geProtoDbroot * dbroot) {
+  keyhole::dbroot::StringEntryProto * translationEntry = dbroot->add_translation_entry();
+  translationEntry->set_string_id(translationEntryId);
+  ++translationEntryId;
+  return translationEntry;
+}
+
+int addKmlUrlAsTranslationEntry(const string & url, geProtoDbroot * dbroot) {
+  keyhole::dbroot::StringEntryProto * translationEntry = addTranslationEntry(dbroot);
+  int id = translationEntry->string_id();
+  translationEntry->set_string_value(url);
+  keyhole::dbroot::NestedFeatureProto * nested = dbroot->add_nested_feature();
+  keyhole::dbroot::StringIdOrValueProto * kmlUrl = nested->mutable_kml_url();
+  kmlUrl->set_string_id(id);
+  return id;
+}
+
+TEST(ReplaceReferencedKmlTest, TranslationEntryReplaceTest) {
+  const string BEFORE_URL_1 = "http://server.name/kmlfile.kmz";
+  const string BEFORE_URL_2 = "http://other.server.name/newfile.kml";
+  const string AFTER_BASE_URL = "https://different.server";
+  const string AFTER_URL_1 = AFTER_BASE_URL + "/dbroot_kml_000.kml";
+  const string AFTER_URL_2 = AFTER_BASE_URL + "/dbroot_kml_001.kml";
+  geProtoDbroot dbroot;
+  
+  translationEntryId = 0;
+  
+  addTranslationEntry(&dbroot);
+  int id1 = addKmlUrlAsTranslationEntry(BEFORE_URL_1, &dbroot);
+  addTranslationEntry(&dbroot);
+  addTranslationEntry(&dbroot);
+  dbroot.add_nested_feature();
+  dbroot.add_nested_feature();
+  dbroot.add_nested_feature();
+  int id2 = addKmlUrlAsTranslationEntry(BEFORE_URL_2, &dbroot);
+  addTranslationEntry(&dbroot);
+
+  ASSERT_TRUE(dbroot.nested_feature(0).has_kml_url());
+  ASSERT_EQ(dbroot.nested_feature(0).kml_url().string_id(), id1);
+  ReplaceReferencedKml(AFTER_BASE_URL, "", false, &dbroot);
+  ASSERT_EQ(dbroot.nested_feature(0).kml_url().string_id(), id1);
```

Comment:
Suggestion: Check the value of the translation entries for `BEFORE_URL_1` and `BEFORE_URL_2`.